### PR TITLE
[poc] socket: mimic MSG_PEEK by read for windows

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: socket: mimic MSG_PEEK by read for windows
Additional Description:
Windows only support level triggered event. The edge triggered event was emulated by the Envoy. But it depends on the envoy code read all the data out of socket buffer, then the `Read` event won't be triggered again. But for listener filter, we need to peek the data, the peek won't consume the data from the socket buffer, the `Read` event will be triggered again, and becomes a busy loop. 
This was found in the PR below:

https://github.com/envoyproxy/envoy/pull/17395#issuecomment-927788120

This PR is trying to mimic the MSG_PEEK by a normal read.

Risk Level: high
Testing:
Docs Changes:
Release Notes:

### CODE REVIEW

File: source/common/network/io_socket_handle_impl.cc [hunk 1]
```diff
@@ -502,6 +512,46 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmmsg(RawSliceArrays& slices, uin
...
+    } else {
+      if (buffer_->length() > 0) {
+        auto copy_size = std::min(buffer_->length(), length);
+        buffer_->copyOut(0, copy_size, buffer);
+        buffer_->drain(copy_size);
+        if (copy_size < length) {
+          length = length - copy_size;
+        } else {
+          if (file_event_) {
+            file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
```
Comment (davinci26) [inline source/common/network/io_socket_handle_impl.cc:548]: In general it is assumed that it is a responsibility of the caller to re-active the event if they expect more data unless the error is WOULD_BLOCK

In the case of tls/proxy inspectors they should be able to determine that the data they currently have are not available to do their work and they should re-enable read. This is something that `connection_impl` does well.

File: source/common/network/io_socket_handle_impl.cc [hunk 2]
```diff
@@ -502,6 +512,46 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmmsg(RawSliceArrays& slices, uin
 }
 
 Api::IoCallUint64Result IoSocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if constexpr (Event::PlatformDefaultTriggerType == Event::FileTriggerType::EmulatedEdge) {
```
Comment (davinci26) [inline source/common/network/io_socket_handle_impl.cc:515]: I am wondering if we should have a dedicated peek function for this.

File: source/common/network/io_socket_handle_impl.cc [hunk 3]
```diff
@@ -114,6 +114,16 @@ Api::IoCallUint64Result IoSocketHandleImpl::read(Buffer::Instance& buffer,
   if (max_length == 0) {
     return Api::ioCallUint64ResultNoError();
   }
+  if constexpr (Event::PlatformDefaultTriggerType == Event::FileTriggerType::EmulatedEdge) {
+    if (buffer_->length() > 0) {
+      auto move_length = std::min(buffer_->length(), max_length);
+      buffer.move(*buffer_);
+      if (file_event_) {
```
Comment (davinci26) [inline source/common/network/io_socket_handle_impl.cc:121]: I am not sure if this is needed. The caller should be reading until `EAGAIN` is returned

**Review by davinci26 — COMMENTED** [review]
some questions I have.

File: source/common/network/io_socket_handle_impl.cc (hunk 1 above)
Comment (soulxu) [reply source/common/network/io_socket_handle_impl.cc:548]: yea, same as the previous one, whether we should emulate all the cases. or we can just add assert here, assume people always read the data until WOULD_BLOCK return.

**Comment by soulxu** [comment]
@davinci26 thanks for the review!

sorry, reply late, I'm in holidays, will back to work next week.

**Comment by soulxu** [comment]
> @soulxu if you are busy let me know and I can take over this PR

thanks! I almost done a version, I should be able to send out today. I should ask your help in the beginning, my windows dev env is on my laptop, and my laptop is too slow~, anyway let me push out this version. if there are more things to do, I will appreciate you can help it.

**Pull request closed without merging**